### PR TITLE
add link identifiers into emitted tags

### PR DESCRIPTION
**Pull request by Gozala**
This is implementation for #434. I do not like the fact that `id` is added for both inline and reference links, however different approaches were way more intrusive (and more difficult for newbie to implement). I have proposed multiple options in the issues maybe we can discuss to identify one that would be most appropriate.

### CODE REVIEW

**Comment by marcusklaas** [comment]
Sorry about the long radio silence. I acknowledge that your use case is a valid one, but I have my doubts about widespread its use is.

Another way you could achieve this goal without having to modify pulldown itself is to use the links source map to read the tag.

**Comment by Gozala** [comment]
> Sorry about the long radio silence. I acknowledge that your use case is a valid one, but I have my doubts about widespread its use is.
 
I do not know. However reference labels are internally parsed and saved to do resolution so it not really introducing complexity merely just exposing information that is already there.

> Another way you could achieve this goal without having to modify pulldown itself is to use the links source map to read the tag.

Perhaps once #430 is fixed that would be an option.

**Comment by Gozala** [comment]
@marcusklaas I think it is also worth considering issue https://github.com/raphlinus/pulldown-cmark/issues/423 in this context. And specifically:

> I think it is worth considering how currently [`Event::FootnoteReference(CowStr<'a>)`](https://docs.rs/pulldown-cmark/0.7.0/pulldown_cmark/enum.Event.html#variant.FootnoteReference) and [`Tag::FootnoteDefinition(CowStr<'a>)`](https://docs.rs/pulldown-cmark/0.7.0/pulldown_cmark/enum.Tag.html#variant.FootnoteDefinition) are decoupled from each other and how one can be present without the other.
>
> That is pretty much what I believe we need for link / image reference and their reference definitions.

It appears to me that it would make a lot more sense to have link / image reference be similar to footnote references. Instead of resolving those and concealing details like reference identifiers. In such design exposing reference identifiers is natural. It also would imply that second pass that I believe is there to do reference resolutions would be unnecessary (at least for use cases that don't require serializing markdown to html).

**Comment by alerque** [comment]
> but I have my doubts about widespread its use is.

I have a use-case for it.

**Comment by dsherret** [comment]
+1 would be great to have this or provide a way in pulldown-cmark to lazily get this information out of the node when providing a range for a link.

**Comment by Martin1887** [comment]
I think that enum variants get too much complicated to have unnamed fields instead of named fields, so a new pull request replicating the functionality will be implemented if this is not changed.

Thanks.

File: src/parse.rs [hunk 1]
```diff
@@ -98,11 +98,13 @@ pub enum Tag<'a> {
     Strong,
     Strikethrough,
 
-    /// A link. The first field is the link type, the second the destination URL and the third is a title.
-    Link(LinkType, CowStr<'a>, CowStr<'a>),
+    /// A link. The first field is the link type, the second the destination URL and the third is a title,
+    /// the fourth is the link identifier.
+    Link(LinkType, CowStr<'a>, CowStr<'a>, CowStr<'a>),
```
Comment (Martin1887) [inline src/parse.rs:103]: Too much complicated variant to have 4 unnamed fields.

File: src/parse.rs [hunk 2]
```diff
@@ -98,11 +98,13 @@ pub enum Tag<'a> {
...
 
-    /// An image. The first field is the link type, the second the destination URL and the third is a title.
-    Image(LinkType, CowStr<'a>, CowStr<'a>),
+    /// An image. The first field is the link type, the second the destination URL and the third is a title,
+    /// the fourth is the link identifier.
+    Image(LinkType, CowStr<'a>, CowStr<'a>, CowStr<'a>),
```
Comment (Martin1887) [inline src/parse.rs:107]: Too much complicated variant to have 4 unnamed fields.

**Review by Martin1887 — COMMENTED** [review]
See comments in the code, too much complicated enum variants to have unnamed fields.

**Pull request merged**